Add configs for reusing specs from environments and enable reuse from included concrete environments by default

kwryankrattiger opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Add option to reuse specs from included concrete environments. The default is to reuse.

kwryankrattiger committed:
Concretizer configs to reuse concrete environments

Add new reuse type "environment".
Add reuse selector to select all concrete specs from an environment.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger commented:
Currently the reuse concrete test is failing. I am able to get a successful concretization using a local environment test and cmake version, but getting the test to work the versions from virtuals is proving to be more finicky.

tldahlgren commented on lib/spack/spack/environment/environment.py:
```diff
@@ -854,6 +854,18 @@ def _exclude_duplicate_runtimes(self, nodes):
+    def __str__(self) -> str:
+        return self.path
+
+    def __nonzero(self) -> bool:
```
Should this be `__nonzero__`?

tldahlgren replied:
Also, shouldn't it be `__bool__` for Python 3?

tldahlgren reviewed (commented):
Very preliminary feedback...

tldahlgren commented:
Including @tgamblin per his off-line request.

tgamblin commented:
@kwryankrattiger can we get the tests passing on this one again?

kwryankrattiger commented:
After talking with @becker33 I think the way `reuse` is defined for concrete includes is going to change. The current implementation applies reuse as a property of the `incude-concrete`. However, this is a bit confusing as it is unclear what should be expected when concretizing with `concretizer:reuse:false`.

To make this simpler to reason about the concretization reuse settings for reusing concrete specs from included environments will move under the `concretizer:reuse` section. TIL, this is a highly configurable section!

Here is an example of some use cases. The main additions here are adding the `environment` type which is either the key to the path/name of the environment or value. This idea could probably be extended to specify reuse from specific build caches as well. For this PR only the environment name/path will be used.

```yaml
spack:
  concretizer:
    # Reuse from this environment (including included concrete) but not elsewhere
    reuse:
      from:
      - type: environment
    # or reuse from only my_env included environment
    reuse:
      from:
      - type:
          environment: my_env
    # or reuse from everywhere
    reuse: true

  include-concrete:
  - my_env
  - my_other_env
  ...
```

CC: @becker33 @tgamblin

kwryankrattiger commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3968,6 +3976,31 @@ def _specs_from_mirror():
+        if not ev.is_env_dir(path) and ev.exists(path):
+            path = ev.root(path)
+
+    if active_env and path in active_env.included_specs_by_hash:
```
I believe this makes sense as the desired behavior.

If we reuse an environment path explicitly:
* And that environment path is included concretely, then reuse the included concrete specs
* And that path is *not* included concretely, then reuse from the concrete environment at that path

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3999,6 +4007,31 @@ def _specs_from_mirror():
+        env = active_env
+
+    if env:
+        return map(lambda x: x[1], env.concretized_specs())
```
```suggestion
        return [concrete for _, concrete in env.concretized_specs()]
```
I think this avoids unnecessary function call overhead and is easier to read.

kwryankrattiger replied:
I thought that map, which in my understanding is just a generator function, would be faster since I wouldn't need to allocate the memory for all of the concretized specs. I originally had it the other way.

I agree that is is a bit more readable as a list.

becker33 replied:
I generally prefer lists because generators are easy to write bugs with, but if you want a generator you can still use this syntax.

`return (concrete for _, concrete in env.concretized_specs())`

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3999,6 +4007,31 @@ def _specs_from_mirror():
+    elif path:
+        env = ev.read(path)
+    else:
+        env = active_env
```
I think this behavior is incorrect -- I thought if we're just reusing from "environment" that meant that we're reusing all concrete specs from included environments regardless of path.

kwryankrattiger replied:
yes, that is what this should do. The path is only if you want to select the subset of concrete specs from an environment, but if the path is not provided it will use all of the `concretized_specs`, which includes all included concrete specs.

becker33 replied:
Got it. Because those are the already concrete specs for this environment. If some of the roots are already concretized, they will be reused too. (which I think is also correct).

Because concrete includes are processed at read time, even though they're only written at concretization time.

That's subtle enough that I think we need a comment explaining how it works.

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,39 @@ def test_env_include_concrete_remove_env():
     assert test2.path not in lockfile_as_dict["include_concrete"].keys()
 
 
+def test_env_include_concrete_reuse():
```
I think we need tests for

1. reuse from all environments, but only from environments
2. reuse True includes from environments
3. reuse from one environment but not from another
4. test that reuse from an included environment that has been changed uses the specs from when the including environment was concretized.

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3966,22 +3967,29 @@ def selected_specs(self) -> List[spack.spec.Spec]:
+    def from_environment(configuration, *, include, exclude, path=None) -> "SpecFilter":
+        factory = functools.partial(_specs_from_environment, path=path)
+        return SpecFilter(
+            factory=factory, is_usable=lambda _: True, include=include, exclude=exclude
```
I think `is_usable` should still be `is_reusable` -- we don't want to reuse a spec that isn't valid for reuse even if it came from an included env. Is there a reason not to apply that filter?

kwryankrattiger replied:
Hmm, yeah I think you are right. I changed that because I wasn't getting reuse at one point but I think the reason was related to the spec list I was passing not the `is_reusable` function.

kwryankrattiger replied:
So I found the reason. The mock tests don't have a runtime environment, so this was failing to consider specs for reuse.

kwryankrattiger committed:
Addressing review comments

* Add more dimensions to the reuse included concrete environments test
* Test changing and reconcretizing the included environment does not
change the combined environment.
* Add a helper to getting an environment from name or path

kwryankrattiger committed:
Add test for environment helper function environment_from_name_or_dir

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4028,6 +4043,12 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                        configuration=self.configuration,
+                        include=[],
+                        exclude=[],
+                        env=ev.active_environment(),
```
```suggestion
                        env=ev.active_environment(),  # includes all concrete includes
```
I think that will help future readers

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4042,7 +4063,26 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                            self.configuration,
+                            include=include,
+                            exclude=exclude,
+                            env=ev.active_environment(),
```
Same as above

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,79 @@ def test_env_include_concrete_remove_env():
+    test1_specs_by_hash = test1._concrete_specs_dict()
+
+    # Set the reuse mode for the environment
+    globals()[f"configure_reuse_{configure_reuse}"](combined)
```
```suggestion
    configure_reuse = globals()[f"configure_reuse_{configure_reuse}"]
    configure_reuse(combined)
```
just for clarity

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,79 @@ def test_env_include_concrete_remove_env():
+    assert not any([s in test1_specs_by_hash_updated for s in comb_specs_by_hash])
+
+
+def test_env_include_concrete_reuse_environment():
```
This test doesn't appear to do anything yet.

kwryankrattiger replied:
yup, I added that before switching the the parameterization. It is redundant now.

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,79 @@ def test_env_include_concrete_remove_env():
+    test1_specs_by_hash_updated = test1._concrete_specs_dict()
+
+    # The new test1 environment should not have any concrete specs in it, the old should
+    assert any([s in test1_specs_by_hash for s in comb_specs_by_hash])
```
```suggestion
    assert all([s in comb_specs_by_hash for s in test1_specs_by_hash])
```
And same for the other set of checks

kwryankrattiger replied:
ah yeah, that is correct, good catch!

becker33 reviewed (changes requested):
I think we need another test (or maybe another parametrization) for testing reusing from an environment that is not included.

This is looking good in general, I think it's getting close

kwryankrattiger committed:
Add test for non-included environment. Fix included env behavior

kwryankrattiger commented:
@becker33 I am thinking about the case of `spack concretize -f` where I want to clear the concretized specs from the environment, but I don't want to re-read the included concrete. Currently there is no way to do that, you must re-read the included concrete environments or reuse the existing concrete specs in the environment.

Any ideas on how to express that? I was thinking about adding an option value force. Maybe something like `-f full` ie. force a re-read of includes and flush all concrete specs. Then the default here would be to preserver included concrete envs.

My thought on this is I maybe be consuming someone else's concrete environment and iterating, but they changed their lock and I didn't want to update yet. The idea of requiring `full` in that case over `full` being the default and not adding a `--keep-included` is I think keeping included concrete is the happy path in most cases.

kwryankrattiger commented:
> I think we need another test (or maybe another parametrization) for testing reusing from an environment that is not included.

I added tests for reuse of non-existent mirror (`from_environment_raise`) and listing an external mirror (`from_environment_external_test`).

I also noticed that we were not correctly distinguishing between included and external environments when reusing the environment by path/name. I added a check and a `_specs_from_environment_included_concrete` to handle that case correctly.

becker33 commented:
> @becker33 I am thinking about the case of `spack concretize -f` where I want to clear the concretized specs from the environment, but I don't want to re-read the included concrete. Currently there is no way to do that, you must re-read the included concrete environments or reuse the existing concrete specs in the environment.
> 
> Any ideas on how to express that? I was thinking about adding an option value force. Maybe something like `-f full` ie. force a re-read of includes and flush all concrete specs. Then the default here would be to preserver included concrete envs.
> 
> My thought on this is I maybe be consuming someone else's concrete environment and iterating, but they changed their lock and I didn't want to update yet. The idea of requiring `full` in that case over `full` being the default and not adding a `--keep-included` is I think keeping included concrete is the happy path in most cases.

I don't think we should do anything to enable reusing without updating the concretizations from the upstream environment. Updating your included environments is part of the contract of `concretize -f`. If we want to change that later, that's an orthogonal PR.

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4042,7 +4089,38 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                if isinstance(source["type"], dict):
+                    env_dir = ev.as_env_dir(source["type"].get("environment"))
+                    active_env = ev.active_environment()
+                    if active_env and env_dir in active_env.included_concrete_envs:
```
```suggestion
                    # if the environment is included as a concrete environment, use the local copy
                    # included concrete environments are only updated at concretization time, and
                    # reuse needs to match the included specs
                    if active_env and env_dir in active_env.included_concrete_envs:                    
```

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4042,7 +4089,38 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                                included_concrete=env_dir,
+                            )
+                        )
+                    else:
```
```suggestion
                    else:
                        # This environment is not included -- we can grab the current specs from its lockfile
```

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4042,7 +4089,38 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                            )
+                        )
+                elif source["type"] == "environment":
+                    self.reuse_sources.append(
```
```suggestion
                    # reusing from the current environment implicitly reuses from all included concrete envs
                    self.reuse_sources.append(
```

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,140 @@ def test_env_include_concrete_remove_env():
+    # This test verifies that concretizing with an included concrete
+    # environment with "concretizer:reuse:true" the included
+    # concrete spec overrides the default with mpi@1.0.
+    print(test1.name)
+    print(test1.concrete_roots())
```
We shouldn't have any print statements in the test once this is ready to merge -- is there something we need to `assert` here or is this just leftover from debugging?

kwryankrattiger replied:
leftover debug, I removed the stuff in `asp.py` but missed it here.

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,140 @@ def test_env_include_concrete_remove_env():
+    combined_env.manifest.configuration.update(
+        {
+            "concretizer": {"reuse": {"from": [{"type": {"environment": "external_test"}}]}},
+            "packages": {"mpich": {"require": ["+debug"]}},
```
What does this requirement do for the test? I don't understand -- the reuse should get `+debug` anyway because that's what's available.

kwryankrattiger replied:
The default is no debug, I am using this to create a new spec in an external environment and make sure that it is picked up when I am only requiring `+debug` in the combined environment. If it failed to reuse, it would select a  3.0 version with debug

kwryankrattiger replied:
The default is `~debug`. I am distinguishing here between the `external_test` version of mpich and the `test1` version that is included concretely.

I will add a comment to that effect here.

kwryankrattiger commented:
> Updating your included environments is part of the contract of concretize -f

I agree with that. But there are cases where I want to keep by current includes, but I have modified other parts of my environment that I want to re-concretize against. I am thinking about a different option which does `-f`, but without overriding includes.

I don't think it needs to be in the PR, just a thought.

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1921,140 @@ def test_env_include_concrete_remove_env():
     assert test2.path not in lockfile_as_dict["include_concrete"].keys()
 
 
+def configure_reuse_true(env):
```
Can you refactor these methods into a single method with an argument for which test it's running? I really don't like the test accessing `globals` with an f-string to figure out which method to call.

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1922,132 @@ def test_env_include_concrete_remove_env():
+
+        # All of the test1 specs (mpich@1) should exist in the combined environment
+        assert all([s in comb_specs_by_hash for s in test1_specs_by_hash])
+        # None of the references specs (using mpich@3) should exist in the combined environment
```
```suggestion
        # None of the references specs (using mpich@3) reuse the specs from test1
        # This tests that the reuse is not happening coincidentally
```

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1922,132 @@ def test_env_include_concrete_remove_env():
+        combined_env.manifest.changed = True
+        combined_env.write()
+
+    return None
```
I think we can clean this up a little bit by using the same interface to the environment in all cases.
```suggestion
    override_env = None
    if reuse_mode == "true":
        config = {"concretizer": {"reuse": True}}
    elif reuse_mode == "from_environment":
        config = {"concretizer": {"reuse": {"from": [{"type": "environment"}]}}}
    elif reuse_mode == "from_environment_test1":
            config = {"concretizer": {"reuse": {"from": [{"type": {"environment": "test1"}}]}}}
    elif reuse_mode == "from_environment_external_test":
        # Create a new environment called external_test that enables the "debug"
        # The default is "~debug"
        env("create", "external_test")
        override_env = ev.read("external_test")
        with override_env:
            add("mpich@1.0 +debug")
        override_env.concretize()
        override_env.write()

        # Reuse from the environment that is not included.
        # Specify the requirement for the debug variant. By default this would concretize to use mpich@3.0
        # but with include concrete the mpich@1.0 +debug version from the "external_test" environment will
        # be used.
        config = {
            "concretizer": {"reuse": {"from": [{"type": {"environment": "external_test"}}]}},
            "packages": {"mpich": {"require": ["+debug"]}},
        }

    elif reuse_mode == "from_environment_raise":
        config = {"concretizer": {"reuse": {"from": [{"type": {"environment": "not-a-real-env"}}]}}}

    combined_env.manifest.configuration.update(config)
    combined_env.manifest.changed = True
    combined_env.write()
    
    return override_env
```

becker33 committed:
bugfix: unify settings for included environments

kwryankrattiger committed:
Update reuse concrete tests to work with properly implemented unify

kwryankrattiger committed:
Fix style checks

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1921,6 +1933,136 @@ def test_env_include_concrete_remove_env():
+        # This tests that the reuse is not happening coincidently
+        assert not any([s in test1_specs_by_hash for s in ref_specs_by_hash])
+
+        # Changing the version of MPICH to 3 in the test env should not change
+        # the combined env.
+        with test1:
+            config("change", "packages:mpich:require:@3")
+            concretize("-f")
+        test1_specs_by_hash_updated = test1.specs_by_hash
+
+        # Concretized combinded again (this should still use the old test1 specs)
+        combined.concretize()
+        comb_specs_by_hash = combined.specs_by_hash
+
+        # Ensure that the mpich used by the mpileaks is the mpich from the originally reused test
+        # environment after updating the included test1 environment.
+        comb_mpileaks_spec = [s for s in comb_specs_by_hash.values() if s.name == "mpileaks"]
+        test1_mpich_spec = [s for s in test1_specs_by_hash.values() if s.name == "mpich"]
+        assert len(comb_mpileaks_spec) == 1
+        assert len(test1_mpich_spec) == 1
+        assert comb_mpileaks_spec[0]["mpich"].dag_hash() == test1_mpich_spec[0].dag_hash()
+
+        # None of the new test1 specs (mpich@3) should exist in the combined environment
+        assert not any([s in comb_specs_by_hash for s in test1_specs_by_hash_updated])
```
I think we can remove this -- this is just testing that when we concretize with no new specs, we don't do anything new. That's already tested by plenty of the other concretization tests and we can simplify this test by omitting it.

If we wanted to test something different with this, we could do the reconcretization of `test1` after creating `combined` but before concretizing. Then we could test that we got the new test1 specs when `from_environment_external_test` but not on the other test cases.

kwryankrattiger replied:
That doesn't seem to work since we re-read the included concrete at concretization time. If I use `-f` it will always pick the new MPICH, which is not what I wanted to test here.

I think this is related to additional functionality I wanted to keep old included specs, clear everything else from the environment. We can follow up for that, I don't need it for the common stacks and I think this is sufficient for most use cases.

kwryankrattiger committed:
Drop redundant part of test

alalazo commented:
Just saw this change on `develop`. Curious why this choice:
```yaml
from:
- type:
     environment: my_env
```
instead of this choice:
```yaml
from:
- type: environment
  name: my_env
```
which could have also applied to buildcaches.

tgamblin commented:
@kwryankrattiger or @becker33 can you please update the description here to be more descriptive?

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4054,7 +4089,46 @@ def __init__(self, configuration: spack.config.Configuration) -> None:
+                    env_dir = ev.as_env_dir(source["type"].get("environment"))
+                    active_env = ev.active_environment()
+                    if active_env and env_dir in active_env.included_concrete_envs:
+                        # If environment is included as a concrete environment, use the local copy
+                        # of specs in the active environment.
+                        # note: included concrete environments are only updated at concretization
+                        #       time, and reuse needs to matchthe included specs.
+                        self.reuse_sources.append(
+                            SpecFilter.from_environment_included_concrete(
+                                self.configuration,
+                                include=include,
+                                exclude=exclude,
+                                env=active_env,
+                                included_concrete=env_dir,
+                            )
+                        )
```
@kwryankrattiger I have two questions on this PR as I'm trying to implement #49707 
1. This feature was never documented. Was that on purpose because it's "experimental" etc. or should we document it?
2. Why do we reuse a possibly outdated version of a lockfile if the environment is among the included environments?

kwryankrattiger replied:
1. So as far as this goes it was more a bug fix than a new feature imo We could probably improve the documentation around this a bit more though, but it wasn't part of the scope of this change.
2. For the case of include concrete part of the feature was to include a "snapshot" of a concrete environment. It is up to the user to say when to update the included concrete lockfile. That should be handled via the `-f` flag.

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -422,6 +426,14 @@ def test_env_install_include_concrete_env(unify, install_mockery, mock_fetch):
     assert test2_roots == combined_included_roots[test2.path]
 
+    mpileaks = combined.specs_by_hash[combined.concretized_order[0]]
+    if unify:
+        assert mpileaks["mpi"].dag_hash() in test1_roots
+        assert mpileaks["libelf"].dag_hash() in test2_roots
+    else:
+        # check that unification is not by accident
+        assert mpileaks["mpi"].dag_hash() not in test1_roots
```
I just found out that these tests are buggy, since they don't reflect what user would see when working with an environment from cli. 

The issue is that to compute reused specs we always rely on the `active_environment`, while here we concretize the `combined` environment outside of:
```python
with combined:
    ...
```
so there is no `active_environment`. If there was, a few assertions on `unify: false` would fail. 

Question is: would it be fine to change the expectations to match the current behavior (i.e. specs from an included environment are "reused" when `unify: false`)? I would say "yes" since it will be more consistent in general to behave like that, but wondering if there are use cases I missed.

kwryankrattiger replied:
The distinction between reuse and unify are important here. `unfiy:false` will disable the strict unification, but if `reuse:true` then we will include existing environment specs (ie. the included concrete specs) as reusable.

It is also possible to configure `unify:false` and _only_ reuse from some of the included concrete environments.

Here you include `foo` and `bar`, but only reuse from `foo` when solving. `bar` specs are just injected into the environment in addition, but they are not to be included in the solve.

```
include_concrete:
    - foo
    - bar
unify: false
reuse:
  from: [
    type: environment
    path: foo
  ]
```
